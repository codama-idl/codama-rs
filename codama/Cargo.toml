[package]
name = "codama"
description = "Generate clients, CLIs, documentation and more from your Solana programs"
version = "0.6.0"
repository = { workspace = true }
edition = { workspace = true }
license = { workspace = true }

[dependencies]
codama-macros = { version = "0.6.0", path = "../codama-macros" }

[target.'cfg(not(target_os = "solana"))'.dependencies]
codama-errors = { version = "0.5.3", path = "../codama-errors" }
codama-korok-plugins = { version = "0.5.3", path = "../codama-korok-plugins" }
codama-korok-visitors = { version = "0.5.3", path = "../codama-korok-visitors" }
codama-koroks = { version = "0.5.3", path = "../codama-koroks" }
codama-nodes = { version = "0.5.3", path = "../codama-nodes" }
codama-stores = { version = "0.5.3", path = "../codama-stores" }
proc-macro2 = "1.0"

[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = [
  'cfg(target_os, values("solana"))',
] }
