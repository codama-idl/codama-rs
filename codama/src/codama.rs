use codama_errors::{CodamaError, CodamaResult};
use codama_korok_plugins::{resolve_plugins, DefaultPlugin, KorokPlugin};
use codama_koroks::RootKorok;
use codama_nodes::{Node, NodeTrait, NodeUnionTrait, RootNode};
use codama_stores::RootStore;
use std::path::Path;

pub struct Codama {
    store: RootStore,
    plugins: Vec<Box<dyn KorokPlugin>>,
    with_default_plugin: bool,
}

impl Codama {
    pub fn new(store: RootStore) -> Self {
        Self {
            store,
            plugins: Vec::new(),
            with_default_plugin: true,
        }
        .add_plugin(DefaultPlugin)
    }

    pub fn load<P: AsRef<Path>>(path: P) -> CodamaResult<Self> {
        Ok(Self::new(RootStore::load(path)?))
    }

    pub fn load_all(paths: &[&Path]) -> CodamaResult<Self> {
        Ok(Self::new(RootStore::load_all(paths)?))
    }

    pub fn hydrate(tt: proc_macro2::TokenStream) -> CodamaResult<Self> {
        Ok(Self::new(RootStore::hydrate(tt)?))
    }

    pub fn without_default_plugin(mut self) -> Self {
        self.with_default_plugin = false;
        self
    }

    pub fn add_plugin<T: KorokPlugin + 'static>(mut self, plugin: T) -> Self {
        self.plugins.push(Box::new(plugin));
        self
    }

    pub fn get_korok(&self) -> CodamaResult<RootKorok> {
        RootKorok::parse(&self.store)
    }

    pub fn get_plugins(&self) -> &[Box<dyn KorokPlugin>] {
        match self.with_default_plugin {
            true => &self.plugins,
            false => &self.plugins[1..],
        }
    }

    pub fn get_visited_korok(&self) -> CodamaResult<RootKorok> {
        let mut korok = self.get_korok()?;
        let run_plugins = resolve_plugins(self.get_plugins());
        run_plugins(&mut korok);
        Ok(korok)
    }

    pub fn get_node(&self) -> CodamaResult<Node> {
        let korok = self.get_visited_korok()?;
        korok.node.clone().ok_or(CodamaError::NodeNotFound)
    }

    pub fn get_idl(&self) -> CodamaResult<RootNode> {
        let node = self.get_node()?;
        match node {
            Node::Root(root) => Ok(root),
            _ => Err(CodamaError::UnexpectedNode {
                expected: "RootNode".to_string(),
                actual: node.kind().to_string(),
            }),
        }
    }

    pub fn get_json_idl(&self) -> CodamaResult<String> {
        self.get_idl()?.to_json()
    }
}
