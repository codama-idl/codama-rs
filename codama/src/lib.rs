pub use codama_macros::*;

// Solana programs only need Codama macros, which expand to
// nothing at compile time. Everything else must be excluded
// to avoid compiling non-Solana-compatible code.

#[cfg(not(target_os = "solana"))]
mod codama;

#[cfg(not(target_os = "solana"))]
pub use {
    codama::*, codama_errors::*, codama_korok_plugins::*, codama_korok_visitors::*,
    codama_koroks::*, codama_nodes::*, codama_stores::*,
};
