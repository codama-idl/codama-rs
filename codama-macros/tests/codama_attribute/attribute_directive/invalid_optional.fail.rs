use codama_macros::codama;

#[codama(account(optional = invalid))]
pub struct Test;

fn main() {}
