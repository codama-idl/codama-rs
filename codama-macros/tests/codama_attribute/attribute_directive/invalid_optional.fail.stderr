error: expected a boolean
 --> tests/codama_attribute/attribute_directive/invalid_optional.fail.rs:3:29
  |
3 | #[codama(account(optional = invalid))]
  |                             ^^^^^^^
